### ISSUE TITLE
How to notifications for changes in the datastore to Netconf client

### ISSUE THREAD

**Issue by ashishsaxenahsc** [CESNET/netopeer2#838]
Hi Devs,

We have use case in our project where we are going to modify the rib tables in the datastore. And any changes in the datastore related to the yang model fields should trigger a notification ,which should be sent to the Netconf client. We have found yang model ietf-routing@2018-03-13.yang that allows to make changes in the rib table. We need to update the fields name, address-family, next-hop ,next-hop-address and destination prefix  in the data model. 

Using the sysrepo API  lyd_new_path , we could update these fields and the changes are reflected in the database.
But we are not able to send the notification to the Netconf client in case of any change in the table.  

We have tried using the example notif_subscribe_example.c , but getting the following error :
[ERR]: Module "ietf-routing" does not define any notifications.
 
Can you please suggest how to allow the netconf server to send the notifications for changes in the datastore to netconf client?

Thanks and Regards,
Ashish

**Comment by michalvasko** [comment]
The simplest solution would be to just use `ietf-netconf-notifications:netconf-config-change` notification that should be generated for every config change. Simply subscribe to it filtering only the config changes relevant for you and it should work. Another solution would be to define your own notification, which you must generate on all the configuration changes that interest you. Then, you would again subscribe to it but require no filter.

**Comment by ashishsaxenahsc** [comment]
Hi Michal,

Thank you for the quick reply.

In our case , we want to get notification for change in fields that are related to state data also. When I am subscribing with ietf-netconf-notifications:netconf-config-change , I am getting notification for change in config data , but it is not helpful in monitoring changes for state data.

We are using yang model  ietf-routing@2018-03-13.yang :
     |  +--rw ribs
     |     +--rw rib* [name]
     |        +--rw name              string
     |        +--rw address-family    identityref
     |        +--ro default-rib?      boolean {multiple-ribs}?
     |        +--ro routes
     |        |  +--ro route*
     |        |     +--ro **route-preference**?          route-preference
     |        |     +--ro **next-hop**
     |        |     |  +--ro (next-hop-options)
     |        |     |     +--:(simple-next-hop)
     |        |     |     |  +--ro outgoing-interface?
     |        |     |     |  |       if:interface-ref
     |        |     |     |  +--ro v4ur:**next-hop-address**?
     |        |     |     |  |       inet:ipv4-address
     |        |     |     |  +--ro v6ur:next-hop-address?
     |        |     |     |          inet:ipv6-address
     |        |     |     +--:(special-next-hop)
     |        |     |     |  +--ro special-next-hop?        enumeration
     |        |     |     +--:(next-hop-list)
     |        |     |        +--ro next-hop-list
     |        |     |           +--ro next-hop*
     |        |     |              +--ro outgoing-interface?
     |        |     |              |       if:interface-ref
     |        |     |              +--ro v4ur:address?
     |        |     |              |       inet:ipv4-address
     |        |     |              +--ro v6ur:address?
     |        |     |                      inet:ipv6-address
     |        |     +--ro source-protocol            identityref
     |        |     +--ro active?                    empty
     |        |     +--ro last-updated?              yang:date-and-time
     |        |     +--ro v4ur:**destination-prefix**?   inet:ipv4-prefix
     |        |     +--ro v6ur:destination-prefix?   inet:ipv6-prefix

And have to update the fields  next-hop ,next-hop-address and destination prefix, which are fields related to state information.

Does the capability of getting notification exists for changes in state data as well ?
In that case, is there any way we can achieve notification for change in state related fields?

Regards,
Ashish

**Comment by michalvasko** [comment]
> Does the capability of getting notification exists for changes in state data as well ?

No, because this is not supported at all in NETCONF. But it is true that [*push* operational data](https://netopeer.liberouter.org/doc/sysrepo/master/html/index.html#oper_data) in *sysrepo* support it but for that you are left only with the custom solution of implementing the notification yourself.

The code would be quite short and simple but to write it you need to have a thorough understanding of *sysrepo* and know exactly what you are doing. I can help you only partly because I just do not have time to explain all this, it is in the documentation. In brief, one of the ways of doing it is writing your notification in YANG, installing the module in *sysrepo*, set correct module change callbacks for all the data that interests you, generate the new notification in the callbacks, and then just subscribe to the notification using a NETCONF client. *netopeer2* should then be sending it.

**Comment by ashishsaxenahsc** [comment]
Hi Michal,
Thanks for clarifying the confusion.
I will try to write our own custom yang model .Can you please give any example for setting callback and generating notification for the changes on the data fields? 
Thanks and Regards,
Ashish

**Comment by michalvasko** [comment]
There are some examples but I do not think they would help you that much. I will mention the specific functions and you can look for their use in `sysrepo/examples`.

In brief, one of the ways of doing it is writing your notification in YANG, installing the module in sysrepo, set correct module change callbacks (`sr_module_change_subscribe`) for all the data that interests you, generate the new notification in the callbacks (`sr_event_notif_send`), and then just subscribe to the notification using a NETCONF client (`subscribe` command in *netopeer2-cli*).
